### PR TITLE
Add a pgvector extension package

### PR DESCRIPTION
**Pull request by msullivan**
Contents go into ext::pgvector.

Some compiler changes were made to get various array <-> vector casts
to work.

The test suite includes tests that the indexes are used.

### CODE REVIEW

File: edb/lib/ext/pgvector.edgeql
Comment (1st1) [inline edb/lib/ext/pgvector.edgeql:41]: My stance is that we should drop this. We don't have `<str>` casts for every format of EdgeDB, we use `to_str()` and `to_datatype()` functions for it. The reason for that is configurability of casting + making sure that we only accept data input compatible with EdgeDB's notion of what's acceptable for the given type.

Given that we have this: `select <pgvector::vector>[1.1, 2.1]` I don't think we need to have string casts at least in 3.0.

Moreover, we don't even know how stable the string casting is in pgvector itself, see https://github.com/pgvector/pgvector/issues/147 as an example.

cc @elvis @msullivan @vpetrovykh

**Review by 1st1 — CHANGES_REQUESTED** [review]

File: edb/lib/ext/pgvector.edgeql
Comment (elprans) [reply edb/lib/ext/pgvector.edgeql:41]: Yes, we don't support `str -> array` casts, and neither should we do here.

**Comment by 1st1** [comment]
Can we make `ext::` optional in our name resolution algo? Basically the idea was to only require it if there's shadowing of user namespace of the extension.

cc @elprans

**Comment by elprans** [comment]
> Can we make `ext::` optional in our name resolution algo?

I don't think it's a good idea.  This basically makes things equivalent to `std`, then what's the point of `ext`?

**Comment by 1st1** [comment]
> I don't think it's a good idea. This basically makes things equivalent to std, then what's the point of ext?

I have an answer but the fact that there's *some* discussion here means that we don't want to hastily implement it anyways. Let's have the discussion some other time.

@msullivan let's just drop the string casts and this is good to go, IMO, unless @elprans wants to do a deeper review.

File: edb/lib/ext/pgvector.edgeql
Comment (msullivan) [reply edb/lib/ext/pgvector.edgeql]: I just pushed extensions constraints and a version pin

File: edb/lib/ext/pgvector.edgeql
Comment (msullivan) [reply edb/lib/ext/pgvector.edgeql:41]: My inclination is to keep the casts, but I don't feel strongly.

Comment (elprans) [reply edb/lib/ext/pgvector.edgeql:41]: We don't have a cast from `str` to `array`, so just from that standpoint I would hold off on adding it for `pgvector`.

Comment (vpetrovykh) [reply edb/lib/ext/pgvector.edgeql:41]: I'm fine with the idea that similar to `array` we skip the direct `str` casts in favour of `to_vector` and `to_str`.

Comment (vpetrovykh) [reply edb/lib/ext/pgvector.edgeql:41]: At the end of the day the more "natural" cast from a literal is still from an array of numbers of some sort. Which we have. And we also have a cast to `array<float32>`.

File: edb/lib/ext/pgvector.edgeql
Comment (elprans) [inline edb/lib/ext/pgvector.edgeql:128]: Hmm, so from pgvector README:

> Note: `<#>` returns the negative inner product since Postgres only supports ASC order index scans on operators

Why are we reverting this?  Have you checked that `ORDER BY inner_product(a, b) DESC` still uses the index?

File: edb/lib/ext/pgvector.edgeql
Comment (msullivan) [reply edb/lib/ext/pgvector.edgeql]: Alright, I removed the `str` casts. Not adding conversion functions for now because `to_str` wouldn't be able to go into `std`

File: edb/lib/ext/pgvector.edgeql
Comment (msullivan) [reply edb/lib/ext/pgvector.edgeql]: No, I was sloppy and only checked euclidean distance.

I generalized the test and sure enough it didn't work, so I turned inner_product into neg_inner_product

File: edb/schema/extensions.py
Comment (elprans) [inline edb/schema/extensions.py:346]: Hm. Why?

File: tests/schemas/dump_v4_default.esdl
Comment (1st1) [inline tests/schemas/dump_v4_default.esdl:20]: Shouldn't we require you to have `using extension pgvector;` like with GraphQL? I think we should.

File: tests/test_edgeql_vector.py
Comment (elprans) [inline tests/test_edgeql_vector.py:349]: Wasn't this fixed by #5629

File: tests/schemas/dump_v4_default.esdl
Comment (msullivan) [reply tests/schemas/dump_v4_default.esdl:20]: We do; it's just configured in the test because test schemas elide the `module default` and so you aren't at the toplevel

File: edb/schema/extensions.py
Comment (msullivan) [reply edb/schema/extensions.py:346]: Because create/deleting extensions does operations that aren't allowed in normal DDL. The create side already had this, but there was something on the delete side that ended up needing it too

File: tests/test_edgeql_vector.py
Comment (msullivan) [reply tests/test_edgeql_vector.py:349]: Yeah, I just haven't rebased this yet

Comment (msullivan) [reply tests/test_edgeql_vector.py]: done

File: tests/schemas/dump_v4_default.esdl
Comment (1st1) [reply tests/schemas/dump_v4_default.esdl:20]: Got it, thanks!

**Comment by msullivan** [comment]
I just force pushed a new history, which I will rebase to land

**Review by elprans — APPROVED** [review]

File: edb/lib/ext/pgvector.edgeql
Comment (1st1) [inline edb/lib/ext/pgvector.edgeql:95]: I don't think we need decimal or bigint/decimal casts -- we don't need them and it's questionable if they're useful at all, as there will be major loss of precision here

**Review by 1st1 — APPROVED** [review]
Approving, but I think we should drop decimal/bigint casts.

File: edb/lib/ext/pgvector.edgeql
Comment (elprans) [reply edb/lib/ext/pgvector.edgeql:95]: We allow `float <> decimal/bigint` casts, though.

Comment (elprans) [reply edb/lib/ext/pgvector.edgeql]: E.g

```
select <array<float32>>[1n, 2n, 3n];
```

works, and I don't see a fundamental difference between `array<float32>` and `pgvector::vector` in terms of casting.

**Pull request merged**
